qbraid_qir/qasm3/visitor.py

```diff
@@ -133,17 +145,38 @@ def entry_point(self) -> str:
     def finalize(self) -> None:
         self._builder.ret(None)
 
-    def _set_scope(self, scope: Scope) -> None:
-        self._scope = scope
+    def _push_scope(self, scope: dict) -> None:
+        self._scope.append(scope)
+
+    def _pop_scope(self) -> None:
+        if len(self._scope) > 0:
+            self._scope.pop()
 
-    def _in_gate(self) -> bool:
-        return self._scope == Scope.GATE
+        raise IndexError("Scope list is empty, can not pop")
+
+    def _get_scope(self) -> dict:
+        if len(self._scope) > 0:
+            return self._scope[-1]
+
+        raise IndexError("No scopes available to get")
```

Comment:
This will work, but will be best to mirror the format used above (after it's updated) even though it's not necessary here, simply for clarity / consistency, e.g.

```python
def _get_scope(self) -> dict:
    if len(self._scope) == 0:
        raise IndexError("No scopes available to get")
    return self._scope[-1]
```